File: clrs/go/insertion_sort.go
package main

import "fmt"

func InsertionSort(arr []int) []int {
	fmt.Println("insetion sort")
	fmt.Println("arr: ", arr)

	for i := 1; i < len(arr); i++ {
		val := arr[i]
		j := i - 1

		// order the previous list until arr[j] is no longer bigger than val
		for j >= 0 && arr[j] > val {
			arr[j+1] = arr[j]
			j--
		}

		// insert into the empty space
		// if moved otherwise leave it as is
		arr[j+1] = val
	}
	fmt.Println("sor: ", arr)

	return arr
}

File: clrs/rust/main/src/insertion_sort.rs
pub fn insertion_sort(arr: &mut [i32]) {
    for i in 1..arr.len() {
        //println!("{}", i);
        let val = arr[i];
        let mut j = i;
        //println!("{}", val);

        while j > 0 && arr[j - 1] > val {
            arr[j] = arr[j - 1];
            j -= 1;
        }
        arr[j] = val;
    }
}
